pkg/gardenlet/controller/shoot/seed_registration_control.go

```diff
@@ -730,3 +740,23 @@ func checkSeedAssociations(ctx context.Context, c client.Client, seedName string
 
 	return nil
 }
+
+func mustEnableVPA(ctx context.Context, c client.Client, shoot *gardencorev1beta1.Shoot) (bool, error) {
+	if !gardencorev1beta1helper.ShootWantsVerticalPodAutoscaler(shoot) {
```

Comment:
If the VPA was just disabled but the shoot is still not reconciled, then both shoot and seed controllers will maintain the VPA till the next shoot reconciliation. Will be there any conflicts or a single cluster can be running with two separate VPA deployments?

Reply:
Good catch! There will most likely be conflicts.

Reply:
Hm, which kind of conflicts? Any suggestions how to resolve them?

Reply:
It there is no vpa deployment in the shoot namespace in the seed, then the seed controller should deploy it in the garden namespace inside the seed. Pretty much similar to the case when VPA on the shoot has been just enabled, but waiting next reconciliation to deploy.

Reply:
> Hm, which kind of conflicts?

Conflicts as in multiple `vpa-updater` instances simultaneously trying to evict Pods for example.

Reply:
Hm, is this a problem? The VPA admission controller is still updating the limits, so are these serious conflicts?

Reply:
I would expect that this will not go well. There will even be multiple webhooks, right?
So I think, we should avoid this if we can.
If you remove the check with `gardencorev1beta1helper.ShootWantsVerticalPodAutoscaler` and only rely on the check if the VPA deployment is found, then it would be sufficient, I think.

Reply:
Ah, there won't be multiple webhooks, the second admission controller will overwrite it probably. So forget that part...

Reply:
I had a quick call with @vpnachev. So indeed, we agreed that this situation should be avoided (especially in the reverse case where shoot VPA is disabled it might happen that the seed does not have any VPA at all for a period of time).

The very easiest solution would be to put the VPA deployment task at the very end of the shoot reconciliation flow. This way, after the flow finishes, the `Seed` object is updated with `10s` and either `.spec.settings.verticalPodAutoscaler.enabled` will be `true` or `false`, but the maximum delay would be `10s`, as said.

A more complex solution requires some kind of synchronization between the seed controller and the shooted seed controller. For example, the `DeployVPA` function could annotate the `Seed` object (in case the shoot is a shooted seed) with information about the VPA deployment status, and then the seed controller can, based on this annotation, determine whether to delete/deploy the VPA.

Opinions?

Reply:
I vote for the current solution and open an enhancement issue with the above proposal.

Reply:
+1 for option 1.

Reply:
TBH, I would prefer fixing it right away as we are on it now. I'd also wait for option 1

Reply:
Sorry, yes option 1 makes sense to be implemented now.
I have no strong opinion for the rest.

Reply:
OK, I've made the changes, please check again :)

Reply:
As discussed on the phone. I think, we should remove this check and only use the check below (`vpa-admission-controller` present?) as an indicator for whether to enable the VPA in the Seed.
This way, we can ensure, that the seed registration controller doesn't enable the Seed VPA after the Shoot VPA was disabled but not yet deleted.

Reply:
You could probably also leverage the Cluster resource here, just like in https://github.com/gardener/gardener/pull/2698#discussion_r468030823.

Reply:
I've removed the check for now, ptal